### ISSUE TITLE
Crash when building a port near waterways

### ISSUE THREAD

**Issue by Niektory** [widelands/widelands#3954]
If you load [this save](https://github.com/widelands/widelands/files/4839573/port_waterways_crash.wgf.zip) the game will crash after a few seconds when the port is completed. The reason seems to be waterways blocking space required for the port dock.

```
Message: adding warehouse for player 5 at (99, 109)
Attempting to setup port without neighboring water (coords:  99x109).

Thread 1 "widelands" received signal SIGSEGV, Segmentation fault.
0x00005555570c9e49 in Widelands::PortDock::get_fleet (this=0x0) at ../src/economy/portdock.h:85
85			return fleet_;
(gdb) bt
#0  0x00005555570c9e49 in Widelands::PortDock::get_fleet() const (this=0x0) at ../src/economy/portdock.h:85
#1  0x00005555570bf12a in Widelands::Warehouse::init(Widelands::EditorGameBase&) (this=0x616000313580, egbase=...)
    at ../src/logic/map_objects/tribes/warehouse.cc:578
#2  0x0000555556f8c417 in Widelands::BuildingDescr::create(Widelands::EditorGameBase&, Widelands::Player*, Widelands::Coords, bool, bool, std::vector<std::pair<unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >) const
    (this=0x617000401500, egbase=..., owner=0x6190006a1380, pos=..., construct=false, loading=false, former_buildings=std::vector of length 1, capacity 1 = {...})
    at ../src/logic/map_objects/tribes/building.cc:245
#3  0x0000555556fb631e in Widelands::ConstructionSite::cleanup(Widelands::EditorGameBase&) (this=0x61600070e080, egbase=...)
    at ../src/logic/map_objects/tribes/constructionsite.cc:283
#4  0x0000555556f72618 in Widelands::MapObject::remove(Widelands::EditorGameBase&) (this=0x61600070e080, egbase=...) at ../src/logic/map_objects/map_object.cc:502
#5  0x0000555556f7269d in Widelands::MapObject::destroy(Widelands::EditorGameBase&) (this=0x61600070e080, egbase=...) at ../src/logic/map_objects/map_object.cc:518
#6  0x0000555556f8eef9 in Widelands::Building::destroy(Widelands::EditorGameBase&) (this=0x61600070e080, egbase=...)
    at ../src/logic/map_objects/tribes/building.cc:546
#7  0x0000555556f6c6d7 in Widelands::CmdDestroyMapObject::execute(Widelands::Game&) (this=0x602001e43650, game=...) at ../src/logic/map_objects/map_object.cc:58
#8  0x000055555742e2e1 in Widelands::CmdQueue::run_queue(int, unsigned int&) (this=0x7fffffff87f8, interval=251, game_time_var=@0x7fffffff7f48: 46500426)
    at ../src/logic/cmd_queue.cc:121
#9  0x00005555567aac85 in Widelands::Game::think() (this=0x7fffffff7f40) at ../src/logic/game.cc:592
#10 0x0000555556c164df in InteractiveBase::think() (this=0x62500064c900) at ../src/wui/interactive_base.cc:700
#11 0x0000555556c9e75d in InteractivePlayer::think() (this=0x62500064c900) at ../src/wui/interactive_player.cc:315
#12 0x0000555556a5b63c in UI::Panel::do_think() (this=0x62500064c900) at ../src/ui_basic/panel.cc:506
#13 0x0000555556a58b57 in UI::Panel::do_run() (this=0x62500064c900) at ../src/ui_basic/panel.cc:195
#14 0x000055555643b812 in UI::Panel::run<UI::Panel::Returncodes>() (this=0x62500064c900) at ../src/ui_basic/panel.h:100
#15 0x00005555567aa410 in Widelands::Game::run(Widelands::Game::StartGameType, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    (this=0x7fffffff7f40, start_game_type=Widelands::Game::Loaded, script_to_run="", replay=false, prefix_for_replays="single_player") at ../src/logic/game.cc:562
#16 0x00005555567a89ed in Widelands::Game::run_load_game(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (this=0x7fffffff7f40, filename="save/405.wgf", script_to_run="") at ../src/logic/game.cc:417
#17 0x00005555564268bc in WLApplication::load_game() (this=0x6120000001c0) at ../src/wlapplication.cc:1423
#18 0x00005555564237bd in WLApplication::mainmenu_singleplayer() (this=0x6120000001c0) at ../src/wlapplication.cc:1252
#19 0x00005555564227d5 in WLApplication::mainmenu() (this=0x6120000001c0) at ../src/wlapplication.cc:1154
#20 0x0000555556412d34 in WLApplication::run() (this=0x6120000001c0) at ../src/wlapplication.cc:493
#21 0x000055555640e62e in main(int, char**) (argc=2, argv=0x7fffffffdf68) at ../src/main.cc:42
```

**Screenshot:**
![shot0017](https://user-images.githubusercontent.com/4470900/85907906-6e248d00-b813-11ea-8d0c-a77846682ae5.png)

**Version:**
 - OS: Manjaro Linux
 - Widelands Version: r24837[75189d8@amazons]

**Comment by Noordfrees** [comment]
The solution would be to either destroy the port on completion (which would look like a bug) or to destroy a nearby waterway to make place for the portdock (hm). Better suggestions welcome…

**Comment by frankystone** [comment]
This is likely the opposite of #3880 ...

> The places where you'd like to build the waterway are blocked by the PortDock. This is an invisible immovable that is automatically created by a port-type warehouse and it's the location where ships have to sail to interact with the port. 

Is this invisible immovable created when the construction of the portdock has finished?

**Comment by hessenfarmer** [comment]
in  my opinion the waterways should be cleared by the port like immovables.  
And this should happen in b21 as this might be a major returning issue.     
Crashes should be analysed deeply and considered for the rc if possible. Cause a crash is simply annoying as there is no savegame created.

**Comment by Noordfrees** [comment]
This crash should happen fairly rarely. There are only a handful of official maps with both seafaring and ferries enabled; of those, most have such short WW length limits that there is no reason to build so many waterways directly before a port out of all places; and even then, there is still a good chance that the player happens to leave one node free where a portdock can be created. I'd rate this bug as a cornercase.

+1 for destroying a waterway in this case, though IMHO a message should be sent to the player if that happens.

**Comment by hessenfarmer** [comment]
Well, it might be a corner case and it might be not. However it seems our winterfreezed version gets tested deeply by Niektory. and it is not the only crash reported. If we postpone all these issues to b22 what is the reason doing all these tests.  As said i can live with not so perfect behaviour or missing functions and everything that allows to continue playing. But known crash issues in a release make me feel uncomfortable. 

so the solution might be clear the waterways in the given radius and implement the message in b22 for example

**Comment by Niektory** [comment]
Yeah, I think we should still consider merging minimal crash fixes such as #3949 and #3953, even if it means a slightly delayed release. Crashes are bad, the fixes are unlikely to make matters *worse*.

(For this issue there's no fix yet though.)

**Comment by Noordfrees** [comment]
> https://www.widelands.org/wiki/ReleasingWidelands/#step-2-winter-time-feature-freeze-the-real-feature-freeze:
> Obviously, new surfacing critical bugs are to be vanquished. Non critical bugs are to be deferred to the next version. What's critical should be discussed on the mailing list if there is no consensus.

I am rather averse to changing the tested release procedure during the release…
Ok, we can still fix crashes with one-line changes like #3949, but even #3953 needs very careful testing first to check it won't have side-effects. And destroying a waterway on port completion is even more likely to be non-trivial and to introduce new bugs, which is the last thing we want to risk during winter freeze…

**Comment by thnih0** [comment]
I played game today and I had game crash with same error in stdout.
So, can we fix this issue in version 21?
